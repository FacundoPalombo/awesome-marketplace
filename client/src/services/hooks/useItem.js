import axios from "axios";
import { useEffect, useState } from "react";

export function useItem(id) {
  const [item, setItem] = useState({});
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);
  useEffect(() => {
    async function getItem(id) {
      setItem({});
      setError(null);
      setLoading(true);
      await axios
        .get(`http://0.0.0.0:3001/api/item/${id}`)
        .then((response) => {
          let priceFormatted = response.data.item.price.toLocaleString("es-AR");
          setLoading(false);
          setItem({ ...response.data.item, ...{ price: priceFormatted } });
        })
        .catch((err) => {
          setLoading(false);
          setError({
            code: err.response.status,
            message: err.response.statusText,
          });
        });
    }
    getItem(id);
  }, [id]);
  return [item, loading, error];
}
