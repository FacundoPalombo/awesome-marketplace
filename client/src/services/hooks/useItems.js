import axios from "axios";
import { useEffect, useState } from "react";

export function useItems(query) {
  const [items, setItems] = useState({});
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);
  useEffect(() => {
    async function getItems(query) {
      setLoading(true);
      const response = await axios.get(
        `http://0.0.0.0:3001/api/items?q=${query}`
      );
      setLoading(false);
      setItems(response.data);
      if (response.status >= 400) {
        setLoading(false);
        setError({ code: response.status, message: response.statusText });
      }
    }
    getItems(query);
  }, [query]);
  return [items, loading, error];
}
