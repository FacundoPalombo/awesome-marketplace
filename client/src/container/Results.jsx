import React from "react";
import { useLocation } from "react-router-dom";
import { useItems } from "services/hooks/useItems";
import { Breadcrumbs } from "common/Breadcrumbs/Breadcrumbs";
import { Result } from "components/Result/Result";
import styles from "./Results.module";

export function Results() {
  const query = new URLSearchParams(useLocation().search).get("q");
  const [items, loading, error] = useItems(query);

  return (
    <section className={styles.container}>
      <div className={styles.breadcrumbsContainer}>
        <nav className={styles.breadcrumbs}>
          {items?.categories?.length !== 0 && !error && (
            <Breadcrumbs categories={items.categories} />
          )}
          {/* TODO: improve this xD */}
          {loading && "loading..."}
          {error && (
            <pre>
              Error {error.code} - {error.message}
            </pre>
          )}
        </nav>
      </div>

      <div className={styles.contentContainer}>
        <main className={styles.content} id="results">
          {items?.items?.map((item, index) => (
            <div key={item.id} className={styles.resultContainer}>
              <Result
                id={item.id}
                thumbnail={item.thumbnail}
                shipping={item.shipping}
                title={item.title}
                price={item.price}
                address={item.address}
              />
              {items?.items?.length - 1 !== index && (
                <hr className={styles.divider} />
              )}
            </div>
          ))}
        </main>
      </div>
    </section>
  );
}
