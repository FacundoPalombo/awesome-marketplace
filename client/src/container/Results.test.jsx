import React from "react";
import mockItems from "utils/tests/fixtures/items";
import { renderWithRouter } from "utils/tests/renderWithRouter";
import { Results } from "./Results";
import { useItems } from "services/hooks/useItems";

jest.mock("services/hooks/useItems", () => ({ useItems: jest.fn() }));

describe("GIVEN Results Component", () => {
  describe("WHEN results was retrieved", () => {
    it("THEN should match to snapshot", () => {
      useItems.mockReturnValue([mockItems, false, null]);
      const { asFragment } = renderWithRouter(<Results />);
      expect(asFragment()).toMatchSnapshot();
    });
    it("THEN should render the loader", () => {
      useItems.mockReturnValue([{ categories: [], items: [] }, true, null]);
      const { asFragment } = renderWithRouter(<Results />);
      expect(asFragment()).toMatchSnapshot();
    });
    xit("THEN should change the url search params", () => {});
  });
});
