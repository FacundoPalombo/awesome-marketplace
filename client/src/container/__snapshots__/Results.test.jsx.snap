// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN Results Component WHEN results was retrieved THEN should match to snapshot 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="breadcrumbsContainer"
    >
      <nav
        class="breadcrumbs"
      >
        <nav
          class="breadcrumbs"
        >
          <span
            class="container"
          >
            <a
              class="breadcrumb"
            >
              Celulares y Teléfonos
            </a>
             &gt; 
          </span>
          <span
            class="container"
          >
            <a
              class="breadcrumb"
            >
              Celulares y Smartphones
            </a>
          </span>
        </nav>
      </nav>
    </div>
    <div
      class="contentContainer"
    >
      <main
        class="content"
        id="results"
      >
        <div
          class="resultContainer"
        >
          <a
            class="link"
            href="/items/:MLA899573282"
          >
            <div
              class="result"
            >
              <div
                class="imageContainer"
              >
                <img
                  alt="Xiaomi Redmi 9 (global) Dual Sim 64 Gb Carbon Grey 4 Gb Ram"
                  class="image"
                  src="http://http2.mlstatic.com/D_818899-MLA43159149904_082020-I.jpg"
                />
              </div>
              <div
                class="description"
              >
                <h3
                  class="price"
                >
                  $ 34600 
                  <img
                    class="freeShipping"
                    srcset="file-mock 1x, file-mock 2x"
                  />
                </h3>
                <h4
                  class="title"
                >
                  Xiaomi Redmi 9 (global) Dual Sim 64 Gb Carbon Grey 4 Gb Ram
                </h4>
              </div>
              <h4
                class="location"
              >
                Córdoba
              </h4>
            </div>
          </a>
          <hr
            class="divider"
          />
        </div>
        <div
          class="resultContainer"
        >
          <a
            class="link"
            href="/items/:MLA900742789"
          >
            <div
              class="result"
            >
              <div
                class="imageContainer"
              >
                <img
                  alt="Xiaomi Redmi Note 8 Dual Sim 64 Gb Neptune Blue 4 Gb Ram"
                  class="image"
                  src="http://http2.mlstatic.com/D_690469-MLA43556746460_092020-I.jpg"
                />
              </div>
              <div
                class="description"
              >
                <h3
                  class="price"
                >
                  $ 35999 
                  <img
                    class="freeShipping"
                    srcset="file-mock 1x, file-mock 2x"
                  />
                </h3>
                <h4
                  class="title"
                >
                  Xiaomi Redmi Note 8 Dual Sim 64 Gb Neptune Blue 4 Gb Ram
                </h4>
              </div>
              <h4
                class="location"
              >
                Balvanera
              </h4>
            </div>
          </a>
          <hr
            class="divider"
          />
        </div>
        <div
          class="resultContainer"
        >
          <a
            class="link"
            href="/items/:MLA885444591"
          >
            <div
              class="result"
            >
              <div
                class="imageContainer"
              >
                <img
                  alt="Xiaomi Redmi Note 9 Dual Sim 128 Gb Verde Bosque 4 Gb Ram"
                  class="image"
                  src="http://http2.mlstatic.com/D_649732-MLA44161684792_112020-I.jpg"
                />
              </div>
              <div
                class="description"
              >
                <h3
                  class="price"
                >
                  $ 43649.1 
                  <img
                    class="freeShipping"
                    srcset="file-mock 1x, file-mock 2x"
                  />
                </h3>
                <h4
                  class="title"
                >
                  Xiaomi Redmi Note 9 Dual Sim 128 Gb Verde Bosque 4 Gb Ram
                </h4>
              </div>
              <h4
                class="location"
              >
                Capital Federal
              </h4>
            </div>
          </a>
          <hr
            class="divider"
          />
        </div>
        <div
          class="resultContainer"
        >
          <a
            class="link"
            href="/items/:MLA912545152"
          >
            <div
              class="result"
            >
              <div
                class="imageContainer"
              >
                <img
                  alt="Xiaomi 9a Dual Sim 32 Gb Gris Granito 2 Gb Ram"
                  class="image"
                  src="http://http2.mlstatic.com/D_955855-MLA43167718739_082020-I.jpg"
                />
              </div>
              <div
                class="description"
              >
                <h3
                  class="price"
                >
                  $ 25700 
                  <img
                    class="freeShipping"
                    srcset="file-mock 1x, file-mock 2x"
                  />
                </h3>
                <h4
                  class="title"
                >
                  Xiaomi 9a Dual Sim 32 Gb Gris Granito 2 Gb Ram
                </h4>
              </div>
              <h4
                class="location"
              >
                Palpala
              </h4>
            </div>
          </a>
          <hr
            class="divider"
          />
        </div>
        <div
          class="resultContainer"
        >
          <a
            class="link"
            href="/items/:MLA880292021"
          >
            <div
              class="result"
            >
              <div
                class="imageContainer"
              >
                <img
                  alt="Xiaomi Note 9 Dual Sim 64 Gb Verde Bosque 3 Gb Ram"
                  class="image"
                  src="http://http2.mlstatic.com/D_878392-MLA44161684781_112020-I.jpg"
                />
              </div>
              <div
                class="description"
              >
                <h3
                  class="price"
                >
                  $ 37600 
                  <img
                    class="freeShipping"
                    srcset="file-mock 1x, file-mock 2x"
                  />
                </h3>
                <h4
                  class="title"
                >
                  Xiaomi Note 9 Dual Sim 64 Gb Verde Bosque 3 Gb Ram
                </h4>
              </div>
              <h4
                class="location"
              >
                Córdoba
              </h4>
            </div>
          </a>
        </div>
      </main>
    </div>
  </section>
</DocumentFragment>
`;

exports[`GIVEN Results Component WHEN results was retrieved THEN should render the loader 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <div
      class="breadcrumbsContainer"
    >
      <nav
        class="breadcrumbs"
      >
        loading...
      </nav>
    </div>
    <div
      class="contentContainer"
    >
      <main
        class="content"
        id="results"
      />
    </div>
  </section>
</DocumentFragment>
`;
